meta-mender-core/conf/layer.conf

```diff
@@ -15,7 +15,6 @@ BBFILE_COLLECTIONS += "mender"
 BBFILE_PATTERN_mender = "^${LAYERDIR}/"
 BBFILE_PRIORITY_mender = "6"
 
-INHERIT += "mender-setup"
```

Comment:
This has an unintended side effect, namely that the `MENDER_FEATURES_ENABLE` and `MENDER_FEATURES_DISABLE` variables stop working. Those are supposed to function independently of `mender-full` & friends, exactly so that you can include only the things you need instead of relying on one of the predefined classes. There is also a lot of error checking in `mender-setup` which doesn't work correctly if this is removed and someone is selecting features one by one.

I think instead we should consider splitting this into "things that change the build", and "things that enable the layer to work". The latter would not influence the build, but would enable the use of `MENDER_FEATURES_ENABLE` and `MENDER_FEATURES_DISABLE`.

Out of curiosity, what exactly did you use to find out how the build changes after adding `meta-mender`? Is this something we can test for?

Reply:
In poky/scripts there is a yocto-check-layer.sh script that does all the checks.

I'll see about creating a separate class that can be automatically included but it's kinda frowned upon to make such changes in layer.conf.

Reply:
Once we get this cleaned up and merged, I would definitely recommend adding CI testing for this.

Reply:
I took a look at `mender-setup.bbclass`. For the most part, it is already supposed to have no effect on the build, other than providing default values for variables, and setting things up so that you can enable them. Same goes for all of the `mender-setup*` files.

I did spot some exceptions though, like:
```
IMAGE_INSTALL_append = " mender-client"
```

This should be changed to:
```
IMAGE_INSTALL_append_mender-client-install = " mender-client"
```

So I think we need to make a choice regarding `mender-setup.bbclass` and the other `mender-setup*` files. I'm proposing one of these:

1. Restrict `mender-setup` to do `MENDER_FEATURES` evaluation, and to set default values for `MENDER_` variables and `mender_` functions that are not expected to be used outside of Mender. This is probably not a huge change, but will require fixing the odd exception here and there, like the one I mentioned above. Downside is that many variables are still evaluated, but they are unlikely to have any direct effect.

2. Restrict `mender-setup` to do `MENDER_FEATURES` evaluation only. This means we would do something like:
   ```
   mender-setup.bbclass:
   ---------------------

   # Code for MENDER_FEATURES evaluation. At minimum it is these three:

   # ------------------------------------------------------------
   DISTRO_FEATURES_BACKFILL_append = " ${MENDER_FEATURES_ENABLE}"
   DISTRO_FEATURES_BACKFILL_CONSIDERED_append = " ${MENDER_FEATURES_DISABLE}"

   python() {
       # Add all possible Mender features here. This list is here to have an
       # ...

   def mender_feature_is_enabled(feature, if_true, if_false, d):
       # ...
   # ------------------------------------------------------------

   # And then, to get the variables, we could do:
   MENDER_SETUP_VARIABLES_CLASS = "${@mender_feature_is_enabled('mender-client-install', 'mender-setup-variables', '', d)}"
   inherit ${MENDER_SETUP_VARIABLES_CLASS}
   ```
   `mender-setup-variables.bbclass` would then contain most of what is in `mender-setup.bbclass` now. I guess this is not a huge amount of work either, mostly just shuffling things around.

3. Don't include `mender-setup` at all, thereby breaking `MENDER_FEATURES` evaluation, and require manual inclusion of `mender-setup` in every build. This has a definite downside, namely that it cannot be added to zeus, as this would break existing builds that select features one by one.

I suppose that number 3 is the cleanest of the three, but it is also the most intrusive. I think I would vote for number 2, as it seems relatively straightforward to implement, is still fairly clean (only `MENDER_FEATURES` are evaluated OOTB), and IMHO provides the best compromise between staying compatible, being clean and providing an easy path to enable features.

Reply:
Is #3 viable for master? As you say it is the cleanest and I don't think an unrealistic expectation.

Reply:
Yes, it is viable for master. Note that we released dunfell yesterday, so strictly speaking it is not viable, but if we act quickly, I suppose we can sneak it into the next tag. It's already slightly broken because of the meta-raspberrypi issue.

Reply:
Actually, looking further, we need at least the mender_feature_is_enabled function automatically added to avoid parse failures.  I _think_ we have to go with #2 because of that.  I'll experiment and post a confirmation later today.